catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultIterable.java

```diff
@@ -0,0 +1,121 @@
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.util.impl;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.apache.commons.lang3.Validate;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Result;
+import ddf.catalog.operation.SourceResponse;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+
+/**
+ * Effectively a cursor over the iteratorResults of a filter that automatically pages through all iteratorResults
+ * <br/>
+ * Throws a {@link CatalogRuntimeException} if anything goes wrong during iteration or
+ * querying
+ */
+public class QueryResultIterable implements Iterable<Result> {
+    private final CatalogFramework catalog;
+
+    private final QueryRequestImpl queryRequest;
+
+    /**
+     * For paging through a single filter with a default BATCH_SIZE of 64
+     *
+     * @param catalog      catalog to query
+     * @param queryRequest A filter to query by
+     */
+    public QueryResultIterable(CatalogFramework catalog, QueryRequestImpl queryRequest) {
+        Validate.notNull(catalog, "Catalog is null");
+        Validate.notNull(queryRequest, "queryRequest is null");
+
+        this.catalog = catalog;
+        this.queryRequest = queryRequest;
+    }
+
+    @Override
+    public Iterator<Result> iterator() {
+        return new ResultQueryIterator();
+    }
+
+    @Override
+    public Spliterator<Result> spliterator() {
+        int characteristics = Spliterator.DISTINCT;
+        return Spliterators.spliteratorUnknownSize(this.iterator(), characteristics);
+    }
+
+    public Stream<Result> stream() {
+        return StreamSupport.stream(this.spliterator(), false);
+    }
+
+    class ResultQueryIterator implements Iterator<Result> {
+        private SourceResponse response = null;
+
+        private QueryResultPaginator queryResultPaginator;
+
+        private Iterator<Result> results = null;
+
+//        private boolean isDone = false;
+
+        public ResultQueryIterator() {
+            this.queryResultPaginator = new QueryResultPaginator(catalog, queryRequest);
+        }
+
+        @Override
+        public boolean hasNext() {
+            ensureInitialized();
+            queryNext();
+
+            if (results.hasNext()) {
+                return true;
+            }
+            // TODO: 6/20/17 Not sure what to do with this clause
+            //            if (isDone) {
+            //                return false;
+            //            }
+            return hasNext();
```

Comment:
Wouldn't this cause an infinite loop when `results.hashNext()` returns false? I think in both cases we should return `results.hasNext()`